x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -60,40 +73,143 @@ public void triggered(SchedulerEngine.Event event) {
+                        // Find all the snapshots that are past their retention date
+                        final List<Tuple<String, SnapshotInfo>> snapshotsToBeDeleted = allSnapshots.stream()
+                            .filter(snapshot -> snapshotEligibleForDeletion(snapshot.v2(), allSnapshots, policiesWithRetention))
+                            .collect(Collectors.toList());
 
-                // Finally, delete the snapshots that need to be deleted
-                deleteSnapshots(snapshotsToBeDeleted);
+                        // Finally, delete the snapshots that need to be deleted
+                        deleteSnapshots(snapshotsToBeDeleted);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        running.set(false);
+                    }
+                }, err -> running.set(false));
 
             } finally {
                 running.set(false);
             }
         } else {
-            logger.debug("snapshot lifecycle retention task started, but a task is already running, skipping");
+            logger.trace("snapshot lifecycle retention task started, but a task is already running, skipping");
         }
     }
 
     static Map<String, SnapshotLifecyclePolicy> getAllPoliciesWithRetentionEnabled(final ClusterState state) {
-        // TODO: fill me in
-        return Collections.emptyMap();
+        final SnapshotLifecycleMetadata snapMeta = state.metaData().custom(SnapshotLifecycleMetadata.TYPE);
+        if (snapMeta == null) {
+            return Collections.emptyMap();
+        }
+        return snapMeta.getSnapshotConfigurations().entrySet().stream()
+            .filter(e -> e.getValue().getPolicy().getRetentionPolicy() != null)
+            .filter(e -> e.getValue().getPolicy().getRetentionPolicy().equals(SnapshotRetentionConfiguration.EMPTY) == false)
+            .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().getPolicy()));
     }
 
-    static boolean snapshotEligibleForDeletion(SnapshotInfo snapshot, Map<String, SnapshotLifecyclePolicy> policies) {
-        // TODO: fill me in
-        return false;
+    static boolean snapshotEligibleForDeletion(SnapshotInfo snapshot, List<Tuple<String, SnapshotInfo>> allSnapshots,
+                                               Map<String, SnapshotLifecyclePolicy> policies) {
+        if (snapshot.userMetadata() == null) {
+            // This snapshot has no metadata, it is not eligible for deletion
+            return false;
+        }
+
+        final String policyId;
+        try {
+            policyId = (String) snapshot.userMetadata().get(SnapshotLifecyclePolicy.POLICY_ID_METADATA_FIELD);
+        } catch (Exception e) {
+            logger.error("unable to retrieve policy id from snapshot metadata [" + snapshot.userMetadata() + "]", e);
+            throw e;
+        }
+
+        SnapshotLifecyclePolicy policy = policies.get(policyId);
```

Comment:
Some implementations of `Map` will throw an NPE if `policyId` is null here (which is possible and not an error condition). Might not be an issue right now but probably best to be safe.